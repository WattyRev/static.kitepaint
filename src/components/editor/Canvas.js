import React from "react";
import styled, { css } from "styled-components";
import PropTypes from "prop-types";
import { appliedColorsShape } from "../../containers/EditorContainer";
import ColorableSvg from "./ColorableSvg";

export const StyleWrapper = styled.div`
  flex-grow: 1;
  ${props => props.theme.patterns.transparencyBackground};
  padding: 16px;
  ${props =>
    props.background
      ? css`
          background: url(${props.background});
          background-size: cover;
          background-position: center;
        `
      : null};

  > div {
    min-height: 50vh;
    width: 100%;
    height: 100%;
    position: relative;
  }
  svg {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    max-width: 100%;
    max-height: 100%;

    *[data-id] {
      cursor: ${props => (props.isReadOnly ? "default" : "pointer")};
    }
  }
`;

/**
 * Canvas is a UI component for the editing area. It displays a product and allows the user to click
 * on enabled panels.
 */
class Canvas extends React.Component {
  static propTypes = {
    background: PropTypes.string,
    hideOutlines: PropTypes.bool,
    /**
     * The product SVG to be rendered and interacted with. Colorable elements (path, polygon, g)
     * should be given a data-id attribute. That is used to indicate that the panel/group can be
     * colored, and it is also used to indicate the coorelation between panels/groups in different
     * variations (used for autofill).
     *
     * A panel/group may also have a data-whitelist property that contains a comma separated list of
     * color names that may be used on that panel/group.
     * A panel/group may also have a data-blacklist property that contains a comma separated list of
     * color names that may not be used on that panel/group.
     */
    svg: PropTypes.string.isRequired,
    /**
     * Called when a colorable panel or group is clicked. Is provided with the data-id value as the
     * first parameter.
     */
    onClick: PropTypes.func,
    /**
     * The currently selected color. This is needed to prevent triggering onClick when a panel/group
     * with data-whitelist is clicked and the current color is not in that whitelist.
     */
    currentColor: PropTypes.string,
    /**
     * A map of colors currently applied to the svg
     */
    colorMap: appliedColorsShape,
    /**
     * Indicates that the canvas is read-only. Will not fire onClick events.
     */
    isReadOnly: PropTypes.bool
  };
  static defaultProps = {
    onClick: () => {},
    colorMap: {},
    currentColor: ""
  };

  /**
   * Handle click events by deciding if it is a relevant item that was clicked.
   * @param  {Object} event The DOM click event
   */
  handleClick = event => {
    // Checks if the the currentColor can be applied given the content or absence of a whitelist.
    const checkWhitelist = whitelist => {
      return (
        !whitelist ||
        !whitelist.length ||
        whitelist.includes(this.props.currentColor.toLowerCase())
      );
    };

    const checkBlacklist = blacklist => {
      return (
        !blacklist ||
        !blacklist.length ||
        !blacklist.includes(this.props.currentColor.toLowerCase())
      );
    };

    // Parses a string form the data-whitelist or data-blacklist property into an array of lowercase,trimmed values
    const processColorList = whitelistString => {
      return whitelistString
        .split(",")
        .map(color => color.trim().toLowerCase())
        .filter(color => !!color);
    };

    // Get the data-id attribute from the target.
    const targetId = event.target.getAttribute("data-id");
    if (targetId) {
      // Get the data-whitelist attribute which contains a comma separated list of color names that can
      // be applied
      const whitelistString = event.target.getAttribute("data-whitelist") || "";
      const whitelist = processColorList(whitelistString);

      // Get the data-blacklist attribute which contains a comma separated list of color names that can
      // be applied
      const blacklistString = event.target.getAttribute("data-blacklist") || "";
      const blacklist = processColorList(blacklistString);

      // If the color satisfies blacklist and whitelist, trigger onClick
      if (checkWhitelist(whitelist) && checkBlacklist(blacklist)) {
        this.props.onClick(targetId);
        return;
      }
      return;
    }

    // If there is no parent element, return
    if (!event.target.parentElement) {
      return;
    }

    // If there was no id on the target, check the parent to see if it is colored as a group
    const parentGroupId = event.target.parentElement.getAttribute("data-id");

    // If there's no group id, then there's nothing to color.
    if (!parentGroupId) {
      return;
    }

    // Get the whitelist of colors for the group
    const whitelistString =
      event.target.parentElement.getAttribute("data-whitelist") || "";
    const whitelist = processColorList(whitelistString);

    // Get the data-blacklist attribute which contains a comma separated list of color names that can
    // be applied
    const blacklistString =
      event.target.parentElement.getAttribute("data-blacklist") || "";
    const blacklist = processColorList(blacklistString);

    // Call onClick if the current color satisfies blacklist and whitelist
    if (checkWhitelist(whitelist) && checkBlacklist(blacklist)) {
      this.props.onClick(parentGroupId);
      return;
    }
  };

  render() {
    return (
      <StyleWrapper
        onClick={this.props.isReadOnly ? () => {} : this.handleClick}
        isReadOnly={this.props.isReadOnly}
        background={this.props.background}
      >
        <ColorableSvg
          hideOutlines={this.props.hideOutlines}
          svg={this.props.svg}
          colorMap={this.props.colorMap}
        />
      </StyleWrapper>
    );
  }
}

export default Canvas;
