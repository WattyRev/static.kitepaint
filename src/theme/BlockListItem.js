import styled from "styled-components";
import { TypographyStyles } from "./Text";

const BlockListItem = styled.div`
  ${TypographyStyles};
  display: block;
  border-bottom: 1px solid
    ${props =>
      props.isLight ? props.theme.colors.grayDark : props.theme.colors.gray};
  color: ${props =>
    props.isLight ? props.theme.colors.silver : props.theme.colors.black};
  padding: 8px;
  text-decoration: none;
  transition: 0.2s background;
`;

export default BlockListItem;
