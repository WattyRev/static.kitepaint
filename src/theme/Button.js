import React from "react";
import styled from "styled-components";
import { TypographyStyles } from "./Text";
import PropTypes from "prop-types";

const StyledButton = styled.button`
  ${TypographyStyles};
  background: linear-gradient(
    -45deg,
    ${props => props.theme.colors.silver},
    ${props => props.theme.colors.silver},
    ${props => props.theme.colors.gray},
    ${props => props.theme.colors.gray}
  );
  background-size: 400% 400%;
  background-position: 100% 50%;
  border-radius: 4px;
  border: 0;
  padding: 4px 8px;
  margin: 8px 0;
  cursor: pointer;
  transition: 0.6s background;
  box-sizing: border-box;
  outline: none;
  text-decoration: none;
  text-align: center;
  color: ${props => props.theme.colors.black};

  &:hover {
    background-position: 0% 50%;
  }
  &.is-primary {
    background-image: linear-gradient(
      -45deg,
      ${props => props.theme.colors.blue},
      ${props => props.theme.colors.blue},
      ${props => props.theme.colors.blueDarker},
      ${props => props.theme.colors.blueDarker}
    );
    color: ${({ theme }) => theme.colors.silver};
  }
  &.is-block {
    display: block;
    width: 100%;
  }
  &[disabled] {
    opacity: 0.5;
    box-shadow: none;
    cursor: not-allowed;
  }
`;

const Button = ({ isPrimary, isBlock, children, className, ...props }) => {
  const classes = className ? [className] : [];
  if (isPrimary) {
    classes.push("is-primary");
  }
  if (isBlock) {
    classes.push("is-block");
  }
  return (
    <StyledButton className={classes.join(" ")} {...props}>
      {children}
    </StyledButton>
  );
};

Button.propTypes = {
  isPrimary: PropTypes.bool,
  isBlock: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string
};

export default Button;
