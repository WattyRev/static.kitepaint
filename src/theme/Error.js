import React from "react";
import styled from "styled-components";
import Icon from "./Icon";
import propTypes from "prop-types";

const StyledError = styled.div`
  color: ${({ theme }) => theme.colors.red};
  border: 1px solid ${({ theme }) => theme.colors.red};
  padding: 4px;
  margin: 0 0 8px;
  border-radius: 4px;
  display: flex;

  > .error-icon {
    padding: 1px 4px 0 0;
  }
`;

/**
 * Displays an error box that can house any type of content.
 */
const Error = ({ children, ...props }) => (
  <StyledError {...props}>
    <Icon className="error-icon" icon="exclamation-circle" />
    <div>{children}</div>
  </StyledError>
);

Error.propTypes = {
  children: propTypes.node
};

export default Error;
