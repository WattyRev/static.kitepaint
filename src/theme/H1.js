import React from "react";
import styled from "styled-components";
import Text from "./Text";
import propTypes from "prop-types";

/**
 * A styled <h1> element.
 * This uses <Text> and passes arguments there.
 * @param {Node} children The items to render inside the <h1>
 */
const H1 = ({ children, ...props }) => (
  <Text as="h1" {...props}>
    {children}
  </Text>
);
H1.propTypes = {
  children: propTypes.node
};
const StyledH1 = styled(H1)`
  font-size: 32px;
  line-height: 32px;
`;
export default StyledH1;
