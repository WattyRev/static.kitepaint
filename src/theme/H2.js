import React from "react";
import styled from "styled-components";
import Text from "./Text";
import propTypes from "prop-types";

/**
 * A styled <h2> element.
 * This uses <Text> and passes arguments there.
 * @param {Node} children The items to render inside the <h2>
 */
const H2 = ({ children, ...props }) => (
  <Text as="h2" {...props}>
    {children}
  </Text>
);
H2.propTypes = {
  children: propTypes.node
};
const StyledH2 = styled(H2)`
  font-size: 24px;
  line-height: 24px;
`;
export default StyledH2;
