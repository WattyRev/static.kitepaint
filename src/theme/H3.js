import React from "react";
import styled from "styled-components";
import Text from "./Text";
import propTypes from "prop-types";

/**
 * A styled <h3> element.
 * This uses <Text> and passes arguments there.
 * @param {Node} children The items to render inside the <h3>
 */
const H3 = ({ children, ...props }) => (
  <Text as="h3" {...props}>
    {children}
  </Text>
);
H3.propTypes = {
  children: propTypes.node
};
const StyledH3 = styled(H3)`
  font-size: 18px;
  line-height: 24px;
`;
export default StyledH3;
