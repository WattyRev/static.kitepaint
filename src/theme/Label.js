import React from "react";
import styled from "styled-components";
import Text from "./Text";
import propTypes from "prop-types";

/**
 * A styled <label> element.
 * This uses <Text> and passes arguments there.
 * @param {Node} children The items to render inside the <label>
 */
const Label = ({ children, ...props }) => (
  <Text as="label" {...props}>
    {children}
  </Text>
);
Label.propTypes = {
  children: propTypes.node
};
const StyledLabel = styled(Label)`
  margin: 8px 0 4px;
  display: block;
  font-weight: 600;
`;
export default StyledLabel;
