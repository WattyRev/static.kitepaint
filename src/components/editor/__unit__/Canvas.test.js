import React from "react";
import { shallow, mount } from "enzyme";
import Theme from "../../../theme";
import Canvas, { StyleWrapper } from "../Canvas";
import { success } from "../../../theme/Alert";

jest.mock("../../../theme/Alert");

function createMockElement(dataId, dataWhitelist, dataBlacklist, parent) {
  return {
    getAttribute: jest.fn(attributeName => {
      const map = {
        "data-id": dataId,
        "data-whitelist": dataWhitelist,
        "data-blacklist": dataBlacklist
      };
      return map[attributeName];
    }),
    parentElement: parent
  };
}

describe("Canvas", () => {
  describe("StyleWrapper", () => {
    it("renders", () => {
      mount(<StyleWrapper theme={Theme} />);
    });
    it("gives cursor: pointer to items with data-id attribute", () => {
      const wrapper = mount(<StyleWrapper theme={Theme} />);
      expect(wrapper).toHaveStyleRule("cursor", "pointer", {
        modifier: "svg *[data-id]"
      });
    });
    it("gives cursor: default to items with data-id attribute if isReadOnly is true", () => {
      const wrapper = mount(<StyleWrapper theme={Theme} isReadOnly />);
      expect(wrapper).toHaveStyleRule("cursor", "default", {
        modifier: "svg *[data-id]"
      });
    });
  });

  let defaultProps;
  beforeEach(() => {
    defaultProps = {
      svg: '<div class="testing_target">test</div>'
    };

    success.mockReset();
  });
  it("renders", () => {
    expect.assertions(1);
    const wrapper = shallow(<Canvas {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe("in readOnly mode", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        isReadOnly: true,
        colorMap: {
          p1: {
            color: "#ffffff",
            name: "white"
          },
          g1: {
            color: "#ff0000",
            name: "red"
          }
        }
      };
    });
    describe("on panel click", () => {
      it("alerts with the color name of the panel", () => {
        expect.assertions(1);
        const event = {
          target: createMockElement("p1", "", "")
        };
        const wrapper = shallow(<Canvas {...defaultProps} />);
        wrapper.find(StyleWrapper).prop("onClick")(event);
        expect(success).toHaveBeenCalledWith("white");
      });
      it("alerts with the color name of the group if it is in a colored group", () => {
        const event = {
          target: createMockElement("", "", "", createMockElement("g1"))
        };
        const wrapper = shallow(<Canvas {...defaultProps} />);
        wrapper.find(StyleWrapper).prop("onClick")(event);
        expect(success).toHaveBeenCalledWith("red");
      });
      it("does not alert if no valid panel was clicked", () => {
        const event = {
          target: createMockElement("", "", "")
        };
        const wrapper = shallow(<Canvas {...defaultProps} />);
        wrapper.find(StyleWrapper).prop("onClick")(event);
        expect(success).not.toHaveBeenCalled();
      });
    });
  });

  describe("when not in readonly mode", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        isReadOnly: false,
        colorMap: {
          p1: {
            color: "#ffffff",
            name: "white"
          },
          g1: {
            color: "#ff0000",
            name: "red"
          }
        }
      };
    });
    describe("on panel click", () => {
      it("does not alert", () => {
        expect.assertions(1);
        const event = {
          target: createMockElement("p1", "", "")
        };
        const wrapper = shallow(<Canvas {...defaultProps} />);
        wrapper.find(StyleWrapper).prop("onClick")(event);
        expect(success).not.toHaveBeenCalled();
      });
    });
  });

  describe("#colorPanel", () => {
    beforeEach(() => {
      defaultProps.onClick = jest.fn();
      defaultProps.currentColor = "orange";
    });
    it("triggers onClick if the clicked element has data-id and no data-whitelist or data-blacklist", () => {
      expect.assertions(2);
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement("p1", "", "")
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).toHaveBeenCalled();
      expect(defaultProps.onClick.mock.calls[0][0]).toEqual("p1");
    });
    it("triggers onClick if the clicked element has data-id and matching data-whitelist", () => {
      expect.assertions(2);
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement(
          "p1",
          // Testing trimming and case insenstivity
          "Blue, white,OranGe , black",
          ""
        )
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).toHaveBeenCalled();
      expect(defaultProps.onClick.mock.calls[0][0]).toEqual("p1");
    });
    it("does not trigger onClick if the clicked element has data-id and matching data-blacklist", () => {
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement(
          "p1",
          "",
          // Testing trimming and case insenstivity
          "Blue, white,OranGe , black"
        )
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).not.toHaveBeenCalled();
    });
    it("triggers onClick if the clicked element has a parent with data-id and no data-whitelist or data-blacklist", () => {
      expect.assertions(2);
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement(
          null,
          null,
          null,
          createMockElement("g1", null, null)
        )
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).toHaveBeenCalled();
      expect(defaultProps.onClick.mock.calls[0][0]).toEqual("g1");
    });
    it("triggers onClick if the clicked element has a parent with data-id and matching data-whitelist", () => {
      expect.assertions(2);
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement(
          null,
          null,
          null,
          createMockElement("g1", "orange, black, red", "")
        )
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).toHaveBeenCalled();
      expect(defaultProps.onClick.mock.calls[0][0]).toEqual("g1");
    });
    it("does not trigger onClick if the clicked element has a parent with data-id and matching data-blacklist", () => {
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement(
          null,
          null,
          null,
          createMockElement("g1", "", "orange, black, red")
        )
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).not.toHaveBeenCalled();
    });
    it("does not trigger onClick if neither the clicked element nor the parent have data-id", () => {
      expect.assertions(1);
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement(null, null, null)
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).not.toHaveBeenCalled();
    });
    it("does not trigger onClick if the clicked element has data-id and a non-matching data-whitelist", () => {
      expect.assertions(1);
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement("p1", "White, Black", "")
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).not.toHaveBeenCalled();
    });
    it("triggers onClick if the clicked element has data-id and a non-matching data-blacklist", () => {
      expect.assertions(1);
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement("p1", "", "White, Black")
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).toHaveBeenCalledWith("p1");
    });
    it("does not trigger onClick if the parent element has data-id and a non-matching data-whitelist", () => {
      expect.assertions(1);
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement(
          null,
          null,
          null,
          createMockElement("g1", "Blue, Green", "")
        )
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).not.toHaveBeenCalled();
    });
    it("triggers onClick if the parent element has data-id and a non-matching data-blacklist", () => {
      expect.assertions(1);
      const subject = new Canvas(defaultProps);
      const event = {
        target: createMockElement(
          null,
          null,
          null,
          createMockElement("g1", "", "Blue, Green")
        )
      };
      subject.colorPanel(event);
      expect(defaultProps.onClick).toHaveBeenCalledWith("g1");
    });
  });
});
