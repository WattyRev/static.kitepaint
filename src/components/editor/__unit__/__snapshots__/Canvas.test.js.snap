// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas renders 1`] = `
<styled.div
  onClick={[Function]}
>
  <svg>
    <defs>
      <filter
        id="valid-panel"
      >
        <feFlood />
        <feComposite
          in2="SourceGraphic"
          operator="out"
        />
        <feMorphology
          operator="dilate"
          radius="4"
        />
        <feGaussianBlur
          stdDeviation="5"
        />
        <feComposite
          in2="SourceGraphic"
          operator="atop"
        />
      </filter>
    </defs>
  </svg>
  <ColorableSvg
    colorMap={Object {}}
    svg="<div class=\\"testing_target\\">test</div>"
  />
</styled.div>
`;
